### PR TITLE
Add codes to solve UnsolvedSymbolException

### PR DESCRIPTION
**Pull request by LoiNguyenCS**
Professor,

This is my attempt to solve our problems with the `UnsolvedSymbolException`.

As we discussed, there are mainly two steps we need to do to prevent `UnsolvedSymbolException`:
1. Figure out the class where the unsolved method belongs to
2. Repeatedly update the synthetic files until there is no UnsolvedSymbolException left

Even though this PR has both steps, I kindly request you to review only the second step, which involves excluding the method `getTheMissingClass` in `UnsolvedSymbolVisitor`.

Right now, we are relying on exception thrown by JavaParser to figure out the class of the unsolved method. But this only works for simple cases like `method.getType()`, `message.toString()`,... This version of `UnsolvedSymbolVisitor` is not able to solve cases like `method.getType().isVoidType()` as `UnsolvedSymbolException` will not tell us the missing class for those cases.

I think my next task in this project is to write a variant of `SymbolSolver` to replace the method `getTheMissingClass` in `UnsolvedSymbolVisitor`, but I think it would be nice if we can refine my codes for the second step first. I think it would reduce the workload for both of us compared to performing code reviews for both steps at once.

### CODE REVIEW

File: src/main/java/org/checkerframework/specimin/SpeciminRunner.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/SpeciminRunner.java:228]: Please be more specific about what might go wrong, beyond "be careful", which is too vague. Ideally, a comment like this should say why a future maintainer needs to be careful. The easiest way to do that is usually to explain what has gone wrong in the past (which is what I suspect happened here), as you do below on line 238.

File: src/main/java/org/checkerframework/specimin/UnsolvedClass.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedClass.java]: Should these fields all be final? They all seem to be set by the constructor, and I don't see any reason they should ever be re-assigned.

File: src/main/java/org/checkerframework/specimin/UnsolvedMethod.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedMethod.java:8]: This isn't enough detail, because of the possibility of overloading (that is, methods that have the same name but different parameter lists). `UnsolvedMethod` should probably use some standard format for this (for example, the [Method Descriptor](https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-4.html#jvms-4.3.3) format used by the JVM)

Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/UnsolvedMethod.java:8]: Whether you choose to use that format or not, `UnsolvedMethod` needs to include at least that much information.

File: src/main/java/org/checkerframework/specimin/UnsolvedMethod.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedMethod.java]: Same comment about these fields being final. We might want to enable a linter like [ErrorProne](https://errorprone.info/) so that you're alerted when you're not following a best practice like this one.

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Typo: "spots out" -> "stubs out"?

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Why is this a `List` rather than a `Set`? Does the order matter?

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:55]: This should say what format the keys are in. You can consult a list of class name formats [here](https://checkerframework.org/manual/#signature-checker).

Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:55]: We should also be running the Signature Checker on this code, if we're going to manipulate classnames as strings (which you're definitely doing in this class).

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Is this a constant? If so, it should be `final` and its name should be `ALL_UPPER_CASE`.

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (kelloggm) [inline src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: In that case, Specimin should stop with a helpful error message (probably by throwing an informative exception).

**Review by kelloggm — CHANGES_REQUESTED** [review]
I have a few high-level concerns:
* we should be running the [Signature Checker](https://checkerframework.org/manual/#signature-checker) to make sure that our use of class name strings is safe/consistent. Any `String` value that holds a classname should have a corresponding annotation as documentation (especially fields).
* fields in the various data classes (`Unsolved*` classes) should be final (and probably private, with getters/setters)
* I don't fully understand the reason that files need to be deleted. I think maybe what's happening is that you're creating stubs for everything used in the target files, then removing whatever isn't used in the target method? If that's the case, the documentation should be explicit about that (and explain why it's better to do it that way rather than only creating what is necessary, which seems like the better design to me).
* related to the above point, I'm concerned about the amount of I/O being done. I expect Specimin should be very fast, but if it needs to create a lot of files and then delete them, it might scale poorly. Ideally, we should only do I/O for things that are actually used by the target method(s).

**Comment by LoiNguyenCS** [comment]
Professor,

> I don't fully understand the reason that files need to be deleted. I think maybe what's happening is that you're creating stubs for everything used in the target files, then removing whatever isn't used in the target method? If that's the case, the documentation should be explicit about that (and explain why it's better to do it that way rather than only creating what is necessary, which seems like the better design to me).

The synthetic files are temporarily created in the root directory of the input so that the input codes can be compilable. So we have to delete all of them (files used by the target method will be included in the output directory, but all the temporary files in the input directory will deleted) since we are not supposed to left anything in the input directory, I think. Another reason is that, I expect the `UnsolvedSymbolVisitor` to run for more than one time in some complicated cases, which means we have to repeatedly update the synthetic files. For those cases, we have to delete the old synthetic files to replace them with the new ones. 

Right now, `UnsolvedSymbolVisitor` is the first phase of Specimin, so at the time `UnsolvedSymbolVisitor` is run, we don't really know the methods that are related to the targeted method.  I think we can update the `UnsolvedSymbolVisitor` so that it is aware of and only takes into consideration classes and methods related to the targeted method. But we still have to delete all of them afterward, and maybe delete them a few times more when we repeatedly update the synthetic files.

**Comment by kelloggm** [comment]
@LoiNguyenCS Thanks for the explanation. I think this is the key point that I did not initially understand, which caused my confusion:
> we don't really know the methods that are related to the targeted method

Because we need to make sure all symbols are resolvable, but we don't have the classpath, and we don't know which parts of the program we're actually going to care about, it is necessary for us to eagerly create files for anything that the target method **might** use. Is that right?

If it is, the design should be clearly documented (probably in the class javadoc on `UnsolvedSymbolVisitor`).

**Comment by LoiNguyenCS** [comment]
Professor,

> Because we need to make sure all symbols are resolvable, but we don't have the classpath, and we don't know which parts of the program we're actually going to care about, it is necessary for us to eagerly create files for anything that the target method might use. Is that right?

Yes, it is right. Alternatively, we can consider somehow running `TargetMethodFinderVisitor` and `UnsolvedSymbolVisitor` at the same time or combine those two visitors as a single visitor. I think that approach will help us avoid creating synthetic files that are not related to the targeted method.  But I am not sure how to do it in detail. I think I can try to working on it after working on a variant of SymbolSolver to replace the method `getTheMissingClass` from `UnsolvedSymbolVisitor`.

**Comment by kelloggm** [comment]
> we can consider somehow running TargetMethodFinderVisitor and UnsolvedSymbolVisitor at the same time or combine those two visitors as a single visitor. I think that approach will help us avoid creating synthetic files that are not related to the targeted method

Let's leave it as a TODO for now, since we're not certain that performance will be a problem.

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (LoiNguyenCS) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Professor, it is not a constant. It is supposed to be replaced with import statements ending with *. As we discussed in our meeting, if a user has a bunch of import statements ending with * and Specimin can't figure out the corresponding import statement for a class, then Specimin will simply pick one import statement ending with * and assume that the unknown class belongs to that import statement. I set its value as `"To.Be.Replaced"` because if I don't give it any value, the compiler will throw an uninitialized field error. But it will definitely be replaced by an import statement ending with *. If there is no import statements ending with *, then this variable just stay there and do nothing.

Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Thank you for the explanation, this makes it a lot clearer.

I'm concerned about what happens if there are more than one * import in a class. Will this string be replaced multiple times? At use sites, you don't really check for that possible problem.

Another thing: rather than the dummy string "To.Be.Replaced", it would be better to use `null` here to represent "no string has been chosen". That will force you to check at use sites, which is desirable. You may also want to check that it is null before you assign it, to avoid its value changing in the middle of processing a class.

Comment (LoiNguyenCS) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Professor, for now, this string will be replaced every time we meet a * import statement during our iteration. So if there are more than  one * import statements, this string will be equal to the last * import statement. Given the fact that Specimin will give a warning to users if they use * import statements, I think it doesn't matter which * import this `chosenPackage` be equal to. This is the piece of code where the string is replaced:

```
  private void setClassAndImportMap() {
    for (String importStatement : this.importStatement) {
      String[] importParts = importStatement.split("\\.");
      if (importParts.length > 0) {
        String className = importParts[importParts.length - 1];
        String packageName = importStatement.replace("." + className, "");
        if (className.equals("*")) {
          chosenPackage = packageName;
        } else {
          this.classAndImportMap.put(className, packageName);
        }
      }
    }
  }
```

Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: > So if there are more than one * import statements, this string will be equal to the last * import statement

This still seems problematic for me. I think it should be an error if this string is non-null when it is assigned (meaning there is a limit of one * import per class), or we should redesign this mechanism so that multiple * imports are supported. Silently getting the package name wrong for some * imports is a problem that will eventually cause us real problems.

Comment (LoiNguyenCS) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: Professor,

I don't understand your concern. Maybe my documentation is misleading. The purpose of this whole thing is simply to know the package directories of imported classes. We do this by analyzing the import statements. For example, if we have these two import statements in the input file:

```
import org.github.sample.Name;
import org.hello.World;
```

Then our `classAndImportMap` will save two keys, which are two classes: `Name` and `World`. The value of each key is the package name in that import statement. So the value of key `Name` is `org.github.sample.Name`, and the value of key `World` is `org.hello.World`.

Thus, if we later realized that we need to create synthetic files for `Name` or `World`, then we can use the `classAndImportMap` to know the package name of that class in order to create the corresponding directory. For example, based on the map, we know that the directory of the synthetic file for class `Name` must be `org/github/sample/Name.java`.

And if our user uses some * import statement, then it is not possible for Specimin to figure out the directory of a missing class. In that case, what Specimin can do is to pick one * import statement randomly (in this case, the last one) and give user a warning.

Comment (kelloggm) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java]: My concern is about this part:
> what Specimin can do is to pick one * import statement randomly (in this case, the last one) and give user a warning

I don't think Specimin should pick one randomly. I think Specimin should actively fail (i.e., throw an exception and stop) if there is ambiguity is which * import to choose. If we do not do this, I am concerned that Specimin will eventually produce synthetic classes that are in the wrong package (because the wrong * import was chosen for them). I don't want that to happen, because I think it will cause problems in some of the use cases I have in mind for Specimin.

File: src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java
Comment (LoiNguyenCS) [reply src/main/java/org/checkerframework/specimin/UnsolvedSymbolVisitor.java:55]: Professor, as we have discussed,  I will resolve this issue in the next PR.

**Comment by LoiNguyenCS** [comment]
Professor, I have updated the codes based on your comments. Please take a look. Thank you.

**Review by kelloggm — APPROVED** [review]
LGTM, assuming there will be two followup PRs:
* one to add the Signature Checker
* one to improve efficiency by avoiding the need to run this visitor over every file in the target project (should only run on files containing targeted methods)

**Pull request merged**
